main.cpp

```diff
@@ -1,75 +1,92 @@
+
+    auto endLoc = Lexer::getLocForEndOfToken(
+        styleCastExpr->getEndLoc(), 0, (*Result.SourceManager), LangOptions());
+    rewriter->InsertText(endLoc, ")");
+  }
 };
 
 class MyASTConsumer : public ASTConsumer {
-public:
-    MyASTConsumer(Rewriter &rewriter) : callback_(rewriter) {
-        matcher_.addMatcher(
-                cStyleCastExpr(unless(isExpansionInSystemHeader())).bind("cast"), &callback_);
-    }
-
-    void HandleTranslationUnit(ASTContext &Context) override {
-        matcher_.matchAST(Context);
-    }
-
-private:
-    CastCallBack callback_;
-    MatchFinder matcher_;
+ public:
+  MyASTConsumer(Rewriter &rewriter) : callback_(rewriter) {
+    matcher_.addMatcher(
+        cStyleCastExpr(unless(isExpansionInSystemHeader())).bind("cast"),
+        &callback_);
+  }
+
+  void HandleTranslationUnit(ASTContext &Context) override {
+    matcher_.matchAST(Context);
+  }
+
+ private:
+  CastCallBack callback_;
+  MatchFinder matcher_;
 };
 
 class CStyleCheckerFrontendAction : public ASTFrontendAction {
-public:
-    CStyleCheckerFrontendAction() = default;
-    
-    void EndSourceFileAction() override {
-        rewriter_.getEditBuffer(rewriter_.getSourceMgr().getMainFileID())
-            .write(llvm::outs());
-    }
-
-    std::unique_ptr<ASTConsumer> CreateASTConsumer(CompilerInstance &CI, StringRef /* file */) override {
-        rewriter_.setSourceMgr(CI.getSourceManager(), CI.getLangOpts());
-        return std::make_unique<MyASTConsumer>(rewriter_);
-    }
-
-private:
-    Rewriter rewriter_;
+ public:
+  CStyleCheckerFrontendAction() = default;
+
+  void EndSourceFileAction() override {
+    rewriter_.getEditBuffer(rewriter_.getSourceMgr().getMainFileID())
+        .write(llvm::outs());
+  }
+
+  std::unique_ptr<ASTConsumer> CreateASTConsumer(
+      CompilerInstance &CI, StringRef /* file */) override {
+    rewriter_.setSourceMgr(CI.getSourceManager(), CI.getLangOpts());
+    return std::make_unique<MyASTConsumer>(rewriter_);
+  }
+
+ private:
+  Rewriter rewriter_;
 };
 
 static llvm::cl::OptionCategory CastMatcherCategory("cast-matcher options");
 
 int main(int argc, const char **argv) {
-    // Ubuntu 20
-    auto Parser = llvm::ExitOnError()(CommonOptionsParser::create(argc, argv, CastMatcherCategory));
-    // For Ubuntu 18, use the old Clang API to construct the CommonOptionsParser:
-    // CommonOptionsParser Parser(argc, argv, CastMatcherCategory);
+  // Ubuntu 20
+  auto Parser = llvm::ExitOnError()(
+      CommonOptionsParser::create(argc, argv, CastMatcherCategory));
+  // For Ubuntu 18, use the old Clang API to construct the CommonOptionsParser:
+  // CommonOptionsParser Parser(argc, argv, CastMatcherCategory);
 
-    ClangTool Tool(Parser.getCompilations(), Parser.getSourcePathList());
-    return Tool.run(newFrontendActionFactory<CStyleCheckerFrontendAction>().get());
+  ClangTool Tool(Parser.getCompilations(), Parser.getSourcePathList());
+  return Tool.run(

```

Comment:
How to prevent changing the C style cast code in macros?
How to prevent parentheses from being added if they are already present in the `(int)(f)` code?

Reply:
1) Для этого можно использовать такую проверку:
```
if (CExpression->getExprLoc().isMacroID()) { return;}
```
2) Оператор "()" является двойным, поэтому достаточно определить наличие хотя бы одной скобки (в данном случае ")":
```
StringRef ref = Lexer::getSourceText(
CharSourceRange::getTokenRange(
styleCastExpr->getRParenLoc().getLocWithOffset(1),
styleCastExpr->getEndLoc()),
*Result.SourceManager, Result.Context->getLangOpts());
std::string sRef = rangeRef.str();

if (range.find(')') == std::string::npos) {
rewriter.InsertText(rParenLoc, ">(");
auto endLoc = Lexer::getLocForEndOfToken(
styleCastExpr.getEndLoc(), 0, (*Result.SourceManager), LangOptions());
rewriter.InsertText(endLoc, ")");
} else {
rewriter.InsertText(rParenLoc, ">");
}

Reply:
Alternate way is:
```cpp
const auto subExpr = styleCastExpr->getSubExprAsWritten();
if (!isa<ParenExpr>(subExpr))
{
    // add parentheses
}
```